src/compiler/scala/tools/nsc/typechecker/Contexts.scala

```diff
@@ -265,6 +265,9 @@ trait Contexts { self: Analyzer =>
     def inSecondTry_=(value: Boolean)         = this(SecondTry) = value
     def inReturnExpr                          = this(ReturnExpr)
     def inTypeConstructorAllowed              = this(TypeConstructorAllowed)
+    def propagateCyclicReferences_=(value: Boolean)
+                                              = this(PropagateCyclicReferences) = value
+    def propagateCyclicReferences             = this(PropagateCyclicReferences)
```

Comment:
I think this is more appropriate as a single boolean for the entire `Global`, rather than one per context. Typechecking in one context can call an type competer in another context which wouldn't have this flag set.

I say that we have another thing called `RecoverableCyclicReference`. Unfortunately, I'm not familar with it. But it seems like it might be something that should be incorporated into this this design. Maybe when the compiler is in `propagate` mode, cyclic errors should be thrown as the recoverable version?